File: src/services/agents/tool-name-normalizer.ts
// src/services/agents/tool-name-normalizer.ts
import { logger } from '@/lib/logger';
import { getAllToolNamesWithCustom } from '@/lib/tools';

/**
 * Validates if a tool name follows the required pattern for AI providers
 * Tool names must match: [a-zA-Z0-9_-]+
 */
export function isValidToolName(toolName: string): boolean {
  const validPattern = /^[a-zA-Z0-9_-]+$/;
  return validPattern.test(toolName);
}

/**
 * Normalizes a tool name by removing invalid characters and mapping to known tool names
 *
 * This function handles cases where AI models return tool names with invalid characters
 * (e.g., "bash Tool" instead of "bash" or "bashTool")
 *
 * @param toolName - The tool name to normalize
 * @returns The normalized tool name, or null if it cannot be mapped to a valid tool
 */
export function normalizeToolName(toolName: string): string | null {
  // Remove all invalid characters (anything not alphanumeric, underscore, or hyphen)
  const cleaned = toolName.replace(/[^a-zA-Z0-9_-]/g, '');

  // If cleaning changed the name, log it
  if (cleaned !== toolName) {
    logger.warn('[ToolNameNormalizer] Invalid tool name detected, attempting to normalize', {
      originalToolName: toolName,
      cleanedName: cleaned,
    });
  }

  // Common mappings from AI-generated names to actual tool names
  // This handles cases like "bash Tool" -> "bash", "Bash Tool" -> "bash", etc.
  const commonMappings: Record<string, string> = {
    // Bash tool variations
    bash: 'bash',
    bashTool: 'bash',
    Bash: 'bash',
    BashTool: 'bash',
    bashtool: 'bash',
    BASH: 'bash',

    // Read file variations
    readFile: 'readFile',
    ReadFile: 'readFile',
    readfile: 'readFile',
    READFILE: 'readFile',
    readFileTool: 'readFile',
    ReadFileTool: 'readFile',

    // Write file variations
    writeFile: 'writeFile',
    WriteFile: 'writeFile',
    writefile: 'writeFile',
    WRITEFILE: 'writeFile',
    writeFileTool: 'writeFile',
    WriteFileTool: 'writeFile',

    // Edit file variations
    editFile: 'editFile',
    EditFile: 'editFile',
    editfile: 'editFile',
    EDITFILE: 'editFile',
    editFileTool: 'editFile',
    EditFileTool: 'editFile',

    // Glob tool variations
    glob: 'glob',
    globTool: 'glob',
    GlobTool: 'glob',
    Glob: 'glob',
    GLOB: 'glob',
    globtool: 'glob',

    // Code search variations
    codeSearch: 'codeSearch',
    CodeSearch: 'codeSearch',
    codesearch: 'codeSearch',
    CODESEARCH: 'codeSearch',
    codeSearchTool: 'codeSearch',
    CodeSearchTool: 'codeSearch',
    GrepTool: 'codeSearch',
    grep: 'codeSearch',
    Grep: 'codeSearch',

    // List files variations
    listFiles: 'listFiles',
    ListFiles: 'listFiles',
    listfiles: 'listFiles',
    LISTFILES: 'listFiles',
    listFilesTool: 'listFiles',
    ListFilesTool: 'listFiles',

    // Call agent variations
    callAgent: 'callAgent',
    CallAgent: 'callAgent',
    callagent: 'callAgent',
    CALLAGENT: 'callAgent',
    callAgentTool: 'callAgent',
    CallAgentTool: 'callAgent',

    // Todo write variations
    todoWrite: 'todoWrite',
    todoWriteTool: 'todoWrite',
    TodoWriteTool: 'todoWrite',
    TodoWrite: 'todoWrite',
    todowrite: 'todoWrite',
    TODOWRITE: 'todoWrite',
    todowritetool: 'todoWrite',

    // Web search variations
    webSearch: 'webSearch',
    webSearchTool: 'webSearch',
    WebSearchTool: 'webSearch',
    WebSearch: 'webSearch',
    websearch: 'webSearch',
    WEBSEARCH: 'webSearch',
    websearchtool: 'webSearch',

    // Web fetch variations
    webFetch: 'webFetch',
    webFetchTool: 'webFetch',
    WebFetchTool: 'webFetch',
    WebFetch: 'webFetch',
    webfetch: 'webFetch',
    WEBFETCH: 'webFetch',
    webfetchtool: 'webFetch',

    // Ask user questions variations
    askUserQuestions: 'askUserQuestions',
    askUserQuestionsTool: 'askUserQuestions',
    AskUserQuestionsTool: 'askUserQuestions',
    AskUserQuestions: 'askUserQuestions',
    askuserquestions: 'askUserQuestions',

    // Memory read variations
    memoryRead: 'memoryRead',
    MemoryRead: 'memoryRead',
    memoryread: 'memoryRead',
    MEMORYREAD: 'memoryRead',
    memory_read: 'memoryRead',
    MemoryReadTool: 'memoryRead',

    // Memory write variations
    memoryWrite: 'memoryWrite',
    MemoryWrite: 'memoryWrite',
    memorywrite: 'memoryWrite',
    MEMORYWRITE: 'memoryWrite',
    memory_write: 'memoryWrite',
    MemoryWriteTool: 'memoryWrite',

    // Exit plan mode variations
    exitPlanMode: 'exitPlanMode',
    exitPlanModeTool: 'exitPlanMode',
    ExitPlanModeTool: 'exitPlanMode',
    ExitPlanMode: 'exitPlanMode',
    exitplanmode: 'exitPlanMode',

    // Get skill variations
    getSkill: 'getSkill',
    getSkillTool: 'getSkill',
    GetSkillTool: 'getSkill',
    GetSkill: 'getSkill',
    getskill: 'getSkill',

    // GitHub PR variations
    githubPR: 'githubPR',
    githubPRTool: 'githubPR',
    GithubPRTool: 'githubPR',
    GithubPR: 'githubPR',
    githubpr: 'githubPR',

    // Test custom tool variations
    test_custom_tool: 'test_custom_tool',
    testCustomTool: 'test_custom_tool',
    TestCustomTool: 'test_custom_tool',
  };

  // Try exact match first
  if (commonMappings[cleaned]) {
    const normalized = commonMappings[cleaned];
    // logger.info('[ToolNameNormalizer] Successfully normalized tool name via mapping', {
    //   originalToolName: toolName,
    //   cleanedName: cleaned,
    //   normalizedName: normalized,
    // });
    return normalized;
  }

  // Try case-insensitive match
  const lowerCleaned = cleaned.toLowerCase();
  for (const [key, value] of Object.entries(commonMappings)) {
    if (key.toLowerCase() === lowerCleaned) {
      // logger.info(
      //   '[ToolNameNormalizer] Successfully normalized tool name via case-insensitive match',
      //   {
      //     originalToolName: toolName,
      //     cleanedName: cleaned,
      //     normalizedName: value,
      //   }
      // );
      return value;
    }
  }

  // If it's an MCP tool (starts with a server ID prefix like "mcp__"), keep the cleaned version
  if (cleaned.startsWith('mcp__') || cleaned.includes('__')) {
    logger.info('[ToolNameNormalizer] Detected MCP tool, using cleaned name', {
      originalToolName: toolName,
      cleanedName: cleaned,
    });
    return cleaned;
  }

  // Check if the cleaned name matches any registered tool names
  const validToolNames = getAllToolNamesWithCustom();
  if (validToolNames.includes(cleaned as any)) {
    logger.info('[ToolNameNormalizer] Cleaned tool name matches a registered tool', {
      originalToolName: toolName,
      cleanedName: cleaned,
    });
    return cleaned;
  }

  logger.error('[ToolNameNormalizer] Unable to normalize tool name to a known tool', {
    originalToolName: toolName,
    cleanedName: cleaned,
    availableTools: validToolNames,
  });

  return null;
}

File: src-tauri/core/src/types/tool_name_normalizer.rs
//! Tool name normalization utilities
//!
//! Normalizes tool identifiers to the canonical camelCase names used by the TS runtime.
//! Accepts legacy snake_case and dash-case aliases for backward compatibility.

use std::collections::HashMap;

/// Canonical tool names used by the TS runtime.
pub const CANONICAL_TOOL_NAMES: &[&str] = &[
    "readFile",
    "writeFile",
    "editFile",
    "glob",
    "codeSearch",
    "listFiles",
    "lsp",
    "bash",
    "webFetch",
    "webSearch",
    "callAgent",
    "todoWrite",
    "askUserQuestions",
    "exitPlanMode",
    "githubPR",
];

/// Normalize a tool name to the canonical camelCase identifier.
///
/// If the name is unknown, returns the trimmed original name.
pub fn normalize_tool_name(name: &str) -> String {
    let trimmed = name.trim();
    if trimmed.is_empty() {
        return String::new();
    }

    let lower = trimmed.to_ascii_lowercase();
    if let Some(canonical) = legacy_aliases().get(lower.as_str()) {
        return canonical.to_string();
    }

    // If already canonical, keep as-is to preserve casing.
    if CANONICAL_TOOL_NAMES.iter().any(|tool| *tool == trimmed) {
        return trimmed.to_string();
    }

    trimmed.to_string()
}

/// Returns true if the tool name is known (canonical or legacy alias).
pub fn is_known_tool_name(name: &str) -> bool {
    let normalized = normalize_tool_name(name);
    CANONICAL_TOOL_NAMES.iter().any(|tool| *tool == normalized)
}

fn legacy_aliases() -> HashMap<&'static str, &'static str> {
    HashMap::from([
        ("read_file", "readFile"),
        ("read-file", "readFile"),
        ("write_file", "writeFile"),
        ("write-file", "writeFile"),
        ("edit_file", "editFile"),
        ("edit-file", "editFile"),
        ("search_files", "codeSearch"),
        ("code_search", "codeSearch"),
        ("code-search", "codeSearch"),
        ("list_files", "listFiles"),
        ("list-files", "listFiles"),
        ("list_directory", "listFiles"),
        ("glob_tool", "glob"),
        ("glob-tool", "glob"),
        ("execute_shell", "bash"),
        ("execute-shell", "bash"),
        ("bash_tool", "bash"),
        ("bash-tool", "bash"),
        ("web_fetch", "webFetch"),
        ("web-fetch", "webFetch"),
        ("web_search", "webSearch"),
        ("web-search", "webSearch"),
        ("call_agent", "callAgent"),
        ("call-agent", "callAgent"),
        ("todo_write", "todoWrite"),
        ("todo-write", "todoWrite"),
        ("ask_user_questions", "askUserQuestions"),
        ("ask-user-questions", "askUserQuestions"),
        ("exit_plan_mode", "exitPlanMode"),
        ("exit-plan-mode", "exitPlanMode"),
        ("github_pr", "githubPR"),
        ("github-pr", "githubPR"),
        ("lsp", "lsp"),
    ])
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn normalizes_aliases_to_canonical() {
        assert_eq!(normalize_tool_name("read_file"), "readFile");
        assert_eq!(normalize_tool_name("write-file"), "writeFile");
        assert_eq!(normalize_tool_name("code_search"), "codeSearch");
        assert_eq!(normalize_tool_name("list_directory"), "listFiles");
        assert_eq!(normalize_tool_name("execute_shell"), "bash");
        assert_eq!(normalize_tool_name("web_fetch"), "webFetch");
        assert_eq!(normalize_tool_name("web_search"), "webSearch");
        assert_eq!(normalize_tool_name("call_agent"), "callAgent");
    }

    #[test]
    fn preserves_canonical_names() {
        assert_eq!(normalize_tool_name("readFile"), "readFile");
        assert_eq!(normalize_tool_name("webFetch"), "webFetch");
        assert_eq!(normalize_tool_name("githubPR"), "githubPR");
    }

    #[test]
    fn unknown_tool_passes_through() {
        assert_eq!(normalize_tool_name("customTool"), "customTool");
    }

    #[test]
    fn empty_name_returns_empty() {
        assert_eq!(normalize_tool_name(""), "");
    }
}
